// @aries-framework/core/build/modules/action-menu/models/ActionMenuOption.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ActionMenuOption = void 0;
const class_transformer_1 = require("class-transformer");
const class_validator_1 = require("class-validator");
const ActionMenuOptionForm_1 = require("./ActionMenuOptionForm");
class ActionMenuOption {
    constructor(options) {
        if (options) {
            this.name = options.name;
            this.title = options.title;
            this.description = options.description;
            this.disabled = options.disabled;
            if (options.form) {
                this.form = new ActionMenuOptionForm_1.ActionMenuForm(options.form);
            }
        }
    }
}
__decorate([
    class_validator_1.IsString(),
    __metadata("design:type", String)
], ActionMenuOption.prototype, "name", void 0);
__decorate([
    class_validator_1.IsString(),
    __metadata("design:type", String)
], ActionMenuOption.prototype, "title", void 0);
__decorate([
    class_validator_1.IsString(),
    __metadata("design:type", String)
], ActionMenuOption.prototype, "description", void 0);
__decorate([
    class_validator_1.IsBoolean(),
    class_validator_1.IsOptional(),
    __metadata("design:type", Boolean)
], ActionMenuOption.prototype, "disabled", void 0);
__decorate([
    class_validator_1.IsInstance(ActionMenuOptionForm_1.ActionMenuForm),
    class_transformer_1.Type(() => ActionMenuOptionForm_1.ActionMenuForm),
    class_validator_1.IsOptional(),
    __metadata("design:type", ActionMenuOptionForm_1.ActionMenuForm)
], ActionMenuOption.prototype, "form", void 0);
exports.ActionMenuOption = ActionMenuOption;
